static/src/javascripts/projects/commercial/modules/article-body-adverts.ts

```diff
@@ -1,66 +1,66 @@
+import type { AdSize } from '@guardian/commercial-core';
+import { adSizes } from '@guardian/commercial-core';
+import { getBreakpoint, getViewport } from 'lib/detect-viewport';
 import config from '../../../lib/config';
-import fastdom from '../../../lib/fastdom-promise';
 import { mediator } from '../../../lib/mediator';
 import { spaceFiller } from '../../common/modules/article/space-filler';
-import { adSizes } from '@guardian/commercial-core';
-import { addSlot } from './dfp/add-slot';
-import { trackAdRender } from './dfp/track-ad-render';
-import { createAdSlot } from './dfp/create-slot';
-
 import { commercialFeatures } from '../../common/modules/commercial/commercial-features';
 import { initCarrot } from './carrot-traffic-driver';
-import { getBreakpoint, getTweakpoint, getViewport } from 'lib/detect-viewport';
+import { addSlot } from './dfp/add-slot';
+import { createAdSlot } from './dfp/create-slot';
+import { trackAdRender } from './dfp/track-ad-render';
 
-const isPaidContent = config.get('page.isPaidContent', false);
+const isPaidContent = config.get<boolean>('page.isPaidContent', false);
 
 const adSlotClassSelectorSizes = {
 	minAbove: 500,
 	minBelow: 500,
 };
 
-const insertAdAtPara = (para, name, type, classes, sizes) => {
+const insertAdAtPara = (
+	para: Node,
+	name: string,
+	type: string,
+	classes?: string,
+	sizes?: Record<string, AdSize[]>,
+): void => {
 	const ad = createAdSlot(type, {
 		name,
 		classes,
 		sizes,
 	});
-
-	return fastdom
-		.mutate(() => {
-			if (para.parentNode) {
-				para.parentNode.insertBefore(ad, para);
-			}
-		})
-		.then(() => {
-			const shouldForceDisplay = ['im', 'carrot'].includes(name);
-			addSlot(ad, shouldForceDisplay);
-		});
```

Comment:
To test but I think that since all `SpacefinderWriter`s are executed as callbacks to `fastdom.mutate` [here](https://github.com/guardian/frontend/pull/24599/files#diff-f04400c4bac9509638e1faa65fe2e9cd92f119f3d4cc027d3a099c04918c99f2R22), the use of fastdom here is redundant.

Reply:
Where we use `fastdom` is not super clear to me. I'd expect it to be as close (i.e. wrapping) to the dom manipulation calls as possible. Otherwise we run the risk of overrunning our frame budget with surrounding logic and missing our nearest animation frame.

Reply:
That's a good point! If a writer is inserting an array of elements we should wrap each insertion in a `mutate` rather than wrapping the entire writer function call.
Maybe the thing to do is to remove `fastdom.mutate` from `spacefiller` and have each writer be responsible for `mutate`-ing. Looks like we'd just have to change `liveblog-adverts.js` and `rich-links.js`